balancer/pickfirst/pickfirst.go

```diff
@@ -103,22 +350,99 @@ func (b *pickfirstBalancer) ResolverError(err error) {
 	})
 }
 
+func (b *pickfirstBalancer) unsetSelectedSubConn() {
+	if b.selectedSubConn != nil {
+		b.selectedSubConn.subConn.Shutdown()
+		b.selectedSubConn = nil
+	}
+}
+
+func (b *pickfirstBalancer) goIdle() {
+	b.unsetSelectedSubConn()
+	b.refreshSunConnList()
```

Comment:
Why do we refresh the subconnList here?

Reply:
Its mainly because the part outside the callback doesn't need synchronization.  If we create the subconns outside and only call connect in the callback, we just need a mutex inside the subConnList. If we create the subConnList in the picker, just before calling connect, we would need a mutex to protect the `balancer.subConnList`.

Let me try to refactor the code to make it easier to create the `subConnList` only when we need to connect.